adding support for proxy configuration to mqtt client

brcourt opened the pull request:
*Issue #, if available:*
#152

*Description of changes:*
Implemented changed to the Paho client.py to allow sockets to be built with proxy support using ~~PySocks~~ a factory function. ~~Changes are based on similar Paho commits.~~ Additionally updated AWSIoTMQTTClient code to implement ~~method~~ factory function configuration.

~~Requires socks package to be imported.~~

~~Lastly, included a commented out example of configuring the client in BasicPubSub.py sample code.~~

Lastly included a new sample called basicPubSubProxy.py

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

bretambrose commented:
Thanks for taking the time to contribute to the SDK.  This is definitely a desired feature, but the implementation appears to add a new third party dependency, which is something we are normally very reluctant to take on.  Is there a way to refactor the implementation to remove the dependency?  Socks might still be the preferred inner implementation in practice, but it would be hidden away from SDK users who didn't need proxy support.

brcourt commented:
Are you suggesting to make the requirement optional for proxy users, or implement proxy support natively to prevent any dependencies at all?

brcourt commented:
Pulled socks.py from PySocks since its the only file thats actually required and added it to the code. Updated setup.py to include new module and importing properly. 

Is this more what you had in mind?

bretambrose commented:
Was thinking something a bit more lightweight.  The end result of all your work is to create a connection in a slightly different manner than normal.  One potential generalization that accommodates that would be to let the user supply a create_connection function.  So if you want to do a socks proxy just bind a lambda to the appropriate proxy_config and pass the function in generically.  It's slightly more work on the app developer's part, but opens up a lot more freedom without dependencies.

We added something similar (client constructor that takes a connection factory) to the Java SDK recently.

brcourt commented:
I disagree, the point of an SDK is to make programming easier. What you just suggested I do is let customers implement a proxy connection entirely on their own. The feature request is to "support proxy config", not to "force users to construct their own sockets" just to support a simple feature.

Further, the code I added does exactly what you say. It uses a custom create_connection function to setup the connection, and the rest is to help negotiate the proxy backend, support HTTP/SOCKS types, or manage slightly different configurations, all of which would be required in the customer's "connection factory". Passing a socket constructor is completely unnecessary and way overkill, especially when this library was explicitly created to solely add proxy support to sockets. 

Lastly, this addition is completely lightweight (for Python). Your suggesting every customer needs to add 800 lines of code and learn socket programming, just to prevent adding 800 lines to the SDK (which would support most all use-cases and require zero knowledge of socket programming to implement). Clearly customers don't want to be forced to write socket connection code to enable proxy.

bretambrose commented:
I think there's a misunderstanding somewhere; nowhere did I suggest adding 800 lines to anything and nowhere did I suggest they write their own socket code.  The lambda they'd pass in, in your use case, would just wrap the socks call.

brcourt commented:
Socks.py does exactly that. And that file is around 800 lines.

bretambrose commented:
My suggestion would change the application developer experience from:

proxy_config = { ... }
client.configProxy(proxy_config)
client.connect()

to

proxy_config = { ... }
connection_factory = lambda ... : socks.create_connection( ... proxy_config ... )
client.configConnectionFactory(connection_factory)
client.connect()

Some work would need to be done within the SDK so that the connection creation calls for connect and reconnect could use the same lambda.

brcourt commented:
Alright I can implement that. It would take the dependency out of the SDK, but customers would still require importing pysocks on their own, or writing their own socket code. Give me a day to make the changes.

brcourt commented:
Proxy configuration is now achievable with custom socket configuration via factory function. Let me know if this isn't what you had in mind.

bretambrose commented on AWSIoTPythonSDK/core/protocol/paho/client.py:
```diff
@@ -780,7 +782,9 @@ def reconnect(self):
         try:
-            if (sys.version_info[0] == 2 and sys.version_info[1] < 7) or (sys.version_info[0] == 3 and sys.version_info[1] < 2):
+            if self._socket_factory:
+                sock = self._socket_factory()
```
I feel like the socket factory lambda should take the same arguments as connect: host, port, timeout, source_address.  You'd pass the appropriate values from the client object.

It can work either way, but if the factory takes no arguments then you have to embed your connect() arguments into it which feels overly coupled.

brcourt replied:
I might misunderstand what you mean, but the socket_factory is not injecting arguments into the socket.create_connection, it's replacing that method entirely. Since we are using a completely different method for creating a socket connection, wouldn't it make sense that all arguments associated with the new factory function be sent as together? If there was another socket library with different arguments for example, wouldn't it block that use case?

bretambrose replied:
You may be right.  I'm going to get another pair of eyes on this and if it's all ok, we'll be merging via an internal repository.  The changes will eventually show up when there's a new version pushed here.

brcourt replied:
Can I get a rough ETA? My company is very interested in having this ability merged as soon as possible.

bretambrose commented:
I will try and get a release out by tomorrow or Monday.

bretambrose commented:
release v1.4.7 contains this PR
